src/xml_parser_main_metadata_together.c

```diff
@@ -337,133 +328,217 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
+            return CR_CB_RET_ERR;
+        } else {
+            // If callback return CRE_OK but it simultaneously set
+            // the tmp_err then it's a programming error.
+            assert(tmp_err == NULL);
+        };
+        // cr_package_free(package)
+    }
+    assert(cr_PkgIterator_is_finished(pkg_iterator));
+
+    cr_PkgIterator_free(pkg_iterator, err);
+
+    if (*err) {
+        return (*err)->code;
+    } else {
+        return CRE_OK;
     }
+}
 
+// TODO: maybe whether or not individual files are parsed could be controlled by NULL paths? I think cr_load_xml_files
+// already works that way.
+cr_PkgIterator *
+cr_PkgIterator_new(const char *primary_path,
+                   const char *filelists_path,
+                   const char *other_path,
+                   cr_XmlParserNewPkgCb newpkgcb,
+                   void *newpkgcb_data,
+                   cr_XmlParserWarningCb warningcb,
+                   void *warningcb_data,
+                   gboolean allow_out_of_order,
+                   GError **err)
+{
     assert(primary_path);
     assert(filelists_path);
     assert(other_path);
-    assert(newpkgcb || pkgcb);
     assert(!err || *err == NULL);
 
-    cr_ParserData *primary_pd = NULL;
-    cr_ParserData *filelists_pd = NULL;
-    cr_ParserData *other_pd = NULL;
+    cr_PkgIterator* new_iter = g_new0(cr_PkgIterator, 1);
+    if (!new_iter) {
+        return NULL;
+    }
+
+    new_iter->primary_path = primary_path;
+    new_iter->filelists_path = filelists_path;
+    new_iter->other_path = other_path;
+
+    new_iter->primary_f = NULL;
+    new_iter->filelists_f = NULL;
+    new_iter->other_f = NULL;
+
+    new_iter->primary_pd = NULL;
+    new_iter->filelists_pd = NULL;
+    new_iter->other_pd = NULL;
 
-    primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    new_iter->primary_is_done = 0;
+    new_iter->filelists_is_done = 0;
+    new_iter->other_is_done = 0;
+
+    cr_CbData *cbdata = g_new0(cr_CbData, 1);
+    new_iter->cbdata = cbdata;
+    cbdata->in_progress_pkgs_list = NULL;
+    cbdata->in_progress_pkgs_hash = NULL;
+    cbdata->finished_pkgs_queue = g_queue_new();
+
+    cbdata->in_progress_count_primary = 0;
+    cbdata->in_progress_count_filelists = 0;
+    cbdata->in_progress_count_other = 0;
+
+    cbdata->newpkgcb = newpkgcb;
+    cbdata->newpkgcb_data = newpkgcb_data;
+
+    new_iter->tmp_err = NULL;
+    new_iter->allow_out_of_order = allow_out_of_order;
+
+    if (allow_out_of_order) {
+        cbdata->in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    }
+
+    GError* tmp_err = new_iter->tmp_err;
+    new_iter->primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", primary_path);
         g_clear_error(&tmp_err);
```

Comment:
This is my mistake but the `g_clear_error(&tmp_err);` shouldn't be here (for all 3 cases of error handling).
The `g_propagate_prefixed_error()` [moves the error](https://docs.gtk.org/glib/func.propagate_prefixed_error.html) so if we later clean it it frees `err` content.